src/runner.cpp

```diff
@@ -0,0 +1,481 @@
+                                       "Runner",
+                                       R"pbdoc(
+Abstract class for derived classes that run an algorithm.
+
+Many of the classes in ``libsemigroups`` implementing the algorithms, that
+are the reason for the existence of this library, are derived from
+:any:`Runner` . The :any:`Runner` class exists to collect various common
+tasks required by such a derived class with a possibly long running
+:any:`run` . These common tasks include:
+
+*  running for a given amount of time (:any:`run_for`)
+*  running until a nullary predicate is true (:any:`run_until`)
+*  checking the status of the algorithm: has it :any:`started` ? :any:`finished`? been killed by another thread (:any:`dead`)? has it timed out (:any:`timed_out`)? has it :any:`stopped` for any reason?
+*  permit the function :any:`run` to be killed from another thread (:any:`kill`).
+)pbdoc");
+
+    py::enum_<Runner::state> state(m, "Runner.state", R"pbdoc(
+Indicates that none of :any:`run` , :any:`run_for` , or :any:`run_until`
+has been called since construction or the last call to :any:`init`.
+)pbdoc");
+    state
+        .value("never_run", Runner::state::never_run, R"pbdoc(
+Indicates that none of :any:`run` , :any:`run_for` , or :any:`run_until`
+has been called since construction or the last call to :any:`init`.
+)pbdoc")
+        .value("running_to_finish", Runner::state::running_to_finish, R"pbdoc(
+Indicates that none of :any:`run` , :any:`run_for` , or :any:`run_until`
+has been called since construction or the last call to :any:`init`.
+)pbdoc")
+        .value("running_for", Runner::state::running_for, R"pbdoc(
+Indicates that none of :any:`run` , :any:`run_for` , or :any:`run_until`
+has been called since construction or the last call to :any:`init`.
+)pbdoc")
+        .value("running_until", Runner::state::running_until, R"pbdoc(
+Indicates that none of :any:`run` , :any:`run_for` , or :any:`run_until`
+has been called since construction or the last call to :any:`init`.
+)pbdoc")
+        .value("timed_out", Runner::state::timed_out, R"pbdoc(
+Indicates that none of :any:`run` , :any:`run_for` , or :any:`run_until`
+has been called since construction or the last call to :any:`init`.
+)pbdoc")
+        .value("stopped_by_predicate",
+               Runner::state::stopped_by_predicate,
+               R"pbdoc(
+Indicates that none of :any:`run` , :any:`run_for` , or :any:`run_until`
+has been called since construction or the last call to :any:`init`.
+)pbdoc")
+        .value("not_running", Runner::state::not_running, R"pbdoc(
+Indicates that none of :any:`run` , :any:`run_for` , or :any:`run_until`
+has been called since construction or the last call to :any:`init`.
+)pbdoc")
+        .value("dead", Runner::state::dead, R"pbdoc(
+Indicates that none of :any:`run` , :any:`run_for` , or :any:`run_until`
+has been called since construction or the last call to :any:`init`.
+)pbdoc");
+    // TODO repr
```

Comment:
Could a repr be something simple like `f"<Runner instance : {self.current_state()}>"`

Reply:
I could add that, but at the same time, it's not possible to create a `Runner` that isn't also something else, so probably I should delete this TODO instead.